File: frontend/omnigo_app/lib/features/customer/data/models/product.dart
class Product {

  Product({
    required this.productTrackingId,
    required this.storeTrackingId,
    required this.name,
    required this.description,
    required this.basePrice,
    this.id = 0,
    this.stock = 0,
    this.imageUrl,
    this.category,
    this.isActive = true,
    this.isFeatured = false,
    this.sku,
    this.storeName,
    this.storeLogoUrl,
    this.storeBannerUrl,
    this.vendorTrackingId,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: (json['id'] as num?)?.toInt() ?? 0,
      productTrackingId: (json['product_tracking_id'] ?? '') as String,
      storeTrackingId: (json['store_tracking_id'] ?? '') as String,
      vendorTrackingId: json['vendor_tracking_id']?.toString(),
      sku: json['sku']?.toString(),
      name: (json['name'] ?? 'Unknown Product') as String,
      description: (json['description'] ?? '') as String,
      basePrice: (json['base_price'] as num?)?.toDouble() ?? 0.0,
      stock: (json['stock'] as num?)?.toInt() ?? 0,
      imageUrl: json['image_url']?.toString(),
      category: json['category']?.toString(),
      isActive: (json['is_active'] as bool?) ?? true,
      isFeatured: (json['is_featured'] as bool?) ?? false,
      storeName: json['store_name']?.toString(),
      storeLogoUrl: json['store_logo_url']?.toString() ?? json['logo_url']?.toString(),
      storeBannerUrl: json['store_banner_url']?.toString() ?? json['banner_url']?.toString(),
    );
  }

  final int id;
  final String productTrackingId;
  final String storeTrackingId;
  final String? vendorTrackingId;
  final String? sku;
  final String name;
  final String description;
  final double basePrice;
  final int stock;
  final String? imageUrl;
  final String? category;
  final bool isActive;
  final bool isFeatured;
  final String? storeName;
  final String? storeLogoUrl;
  final String? storeBannerUrl;

  List<String> get allImages {
    final list = <String>[];
    if (imageUrl != null && imageUrl!.isNotEmpty) {
      list.add(imageUrl!);
    }
    return list;
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'product_tracking_id': productTrackingId,
        'store_tracking_id': storeTrackingId,
        'vendor_tracking_id': vendorTrackingId,
        'sku': sku,
        'name': name,
        'description': description,
        'base_price': basePrice,
        'stock': stock,
        'image_url': imageUrl,
        'category': category,
        'is_active': isActive,
        'is_featured': isFeatured,
        'store_name': storeName,
        'store_logo_url': storeLogoUrl,
        'store_banner_url': storeBannerUrl,
      };
}

File: backend/go-services/internal/product/models/product.go
package models

import "time"

// Product represents an item listed by a vendor store
type Product struct {
	ID                int       `json:"id"`
	ProductTrackingID string    `json:"product_tracking_id"`
	VendorTrackingID  string    `json:"vendor_tracking_id"`
	StoreTrackingID   string    `json:"store_tracking_id"`
	SKU               string    `json:"sku"`
	Name              string    `json:"name"`
	Description       string    `json:"description,omitempty"`
	BasePrice         float64   `json:"base_price"`
	Stock             int       `json:"stock"`
	IsFeatured        bool      `json:"is_featured"`
	ImageURL          string    `json:"image_url"`
	Category          string    `json:"category"`
	IsActive          bool      `json:"is_active"`
	CreatedAt         time.Time `json:"created_at"`
	UpdatedAt         time.Time `json:"updated_at"`
}

// CreateProductRequest is the payload for creating a new product
type CreateProductRequest struct {
	VendorTrackingID string  `json:"vendor_tracking_id" binding:"required"`
	StoreTrackingID  string  `json:"store_tracking_id" binding:"required"`
	SKU              string  `json:"sku"`
	Name             string  `json:"name" binding:"required"`
	Description      string  `json:"description"`
	BasePrice        float64 `json:"base_price" binding:"required"`
	Stock            int     `json:"stock" binding:"required,min=0"`
	ImageURL         string  `json:"image_url"`
	Category         string  `json:"category"`
}

// OrderItem represents a line item for stock reservation
type OrderItem struct {
	ProductTrackingID string  `json:"product_tracking_id"`
	Quantity          int     `json:"quantity"`
	PriceAtCheckout   float64 `json:"price_at_checkout"`  // Returned by reservation
	StoreTrackingID   string  `json:"store_tracking_id"`  // Returned by reservation
	VendorTrackingID  string  `json:"vendor_tracking_id"` // Returned by reservation
}

// VendorCreateProductRequest is the vendor-authenticated variant of
// CreateProductRequest: vendor_tracking_id is pulled from the
// Authorization header (not the JSON body) so a vendor cannot spoof
// another merchant's catalog.
type VendorCreateProductRequest struct {
	StoreTrackingID string  `json:"store_tracking_id" binding:"required"`
	SKU             string  `json:"sku"`
	Name            string  `json:"name" binding:"required"`
	Description     string  `json:"description"`
	BasePrice       float64 `json:"base_price" binding:"required"`
	Stock           int     `json:"stock" binding:"required"`
	ImageURL        string  `json:"image_url"`
	Category        string  `json:"category"`
}

// UpdateProductRequest is the partial-update payload. All fields are
// pointers so callers can send only the subset they want changed.
type UpdateProductRequest struct {
	Name        *string  `json:"name"`
	Description *string  `json:"description"`
	BasePrice   *float64 `json:"base_price"`
	Stock       *int     `json:"stock"`
	ImageURL    *string  `json:"image_url"`
	Category    *string  `json:"category"`
	IsFeatured  *bool    `json:"is_featured"`
	IsActive    *bool    `json:"is_active"`
}
